File: src/smooth/_proxy.py
"""FRP-based proxy tunnel for exposing local SOCKS5 proxy.

This module provides functionality to:
1. Download and install the FRP client binary (frpc)
2. Start a SOCKS5 proxy tunnel to a remote FRP server
3. Manage proxy lifecycle per session

The proxy connects to a remote FRP server and exposes a local SOCKS5 proxy
that can be used by the browser session.
"""

import platform
import shutil
import subprocess
import tarfile
import tempfile
import threading
import urllib.request
import zipfile
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

# FRP version to use
FRP_VERSION = "0.66.0"

# Directory to store FRP binaries and configs
FRP_DIR = Path.home() / ".smooth" / "frp"


@dataclass
class ProxyConfig:
  """Configuration for the FRP proxy tunnel."""

  server_url: str
  token: str
  remote_port: int = 1080
  session_id: str = "default"


@dataclass
class _ProxyState:
  """Internal state for a proxy instance."""

  process: subprocess.Popen[bytes] | None = None
  config_file: Path | None = None
  lock: threading.Lock = field(default_factory=threading.Lock)


class FRPProxy:
  """FRP-based proxy tunnel manager."""

  def __init__(self, config: ProxyConfig):
    """Initialize the FRP proxy.

    Args:
        config: Proxy configuration with server details and token.
    """
    self.config = config
    self._state = _ProxyState()
    self._bin_path: Path | None = None

  @staticmethod
  def _get_platform_info() -> tuple[str, str, str]:
    """Get platform-specific information for FRP download.

    Returns:
        Tuple of (os_name, arch, extension).

    Raises:
        RuntimeError: If the platform is not supported.
    """
    system = platform.system().lower()
    machine = platform.machine().lower()

    # Map OS
    if system == "darwin":
      os_name = "darwin"
    elif system == "windows":
      os_name = "windows"
    else:
      os_name = "linux"

    # Map architecture
    if machine in ["x86_64", "amd64"]:
      arch = "amd64"
    elif machine in ["aarch64", "arm64"]:
      arch = "arm64"
    else:
      raise RuntimeError(f"Unsupported architecture: {machine}")

    ext = "zip" if system == "windows" else "tar.gz"

    return os_name, arch, ext

  def _install_frp(self) -> Path:
    """Download and install FRP binary if not already present.

    Returns:
        Path to the frpc binary.

    Raises:
        RuntimeError: If installation fails.
    """
    FRP_DIR.mkdir(parents=True, exist_ok=True)

    os_name, arch, ext = self._get_platform_info()
    bin_name = "frpc.exe" if os_name == "windows" else "frpc"
    bin_path = FRP_DIR / bin_name

    # Check if binary already exists
    if bin_path.exists():
      return bin_path

    # Construct download URL
    folder_name = f"frp_{FRP_VERSION}_{os_name}_{arch}"
    filename = f"{folder_name}.{ext}"
    url = f"https://github.com/fatedier/frp/releases/download/v{FRP_VERSION}/{filename}"

    try:
      # Download to temp file
      with tempfile.NamedTemporaryFile(delete=False, suffix=f".{ext}") as tmp:
        tmp_path = Path(tmp.name)

      urllib.request.urlretrieve(url, tmp_path)

      # Extract
      extract_dir = FRP_DIR / "extract_tmp"
      extract_dir.mkdir(exist_ok=True)

      if ext == "zip":
        with zipfile.ZipFile(tmp_path, "r") as z:
          z.extractall(extract_dir)
      else:
        with tarfile.open(tmp_path, "r:gz") as t:
          t.extractall(extract_dir)

      # Move binary
      src = extract_dir / folder_name / bin_name
      if bin_path.exists():
        bin_path.unlink()
      shutil.move(str(src), str(bin_path))

      # Cleanup
      tmp_path.unlink()
      shutil.rmtree(extract_dir, ignore_errors=True)

      # Make executable on Unix
      if os_name != "windows":
        bin_path.chmod(0o755)

      return bin_path

    except Exception as e:
      raise RuntimeError(f"Failed to install FRP: {e}") from e

  def _create_config(self) -> Path:
    """Create FRP client configuration file.

    Returns:
        Path to the configuration file.
    """
    FRP_DIR.mkdir(parents=True, exist_ok=True)

    config_path = FRP_DIR / f"frpc_{self.config.session_id}.yml"
    # port should be changed when we use load balancing
    yaml_content = f"""
serverAddr: {self.config.server_url}
serverPort: 7000
auth:
  method: token
  token: "{self.config.token}"

transport:
  protocol: "websocket"
  tls:
    enable: true
    serverName: "{self.config.server_url}"

proxies:
  - name: "socks5_tunnel_{self.config.session_id}"
    type: "tcp"
    remotePort: {self.config.remote_port}
    plugin:
      type: "socks5"
"""
    config_path.write_text(yaml_content)
    return config_path

  def start(self) -> None:
    """Start the FRP proxy tunnel.

    Raises:
        RuntimeError: If the proxy fails to start or is already running.
    """
    with self._state.lock:
      if self._state.process is not None:
        raise RuntimeError("Proxy is already running")

      try:
        # Install FRP if needed
        self._bin_path = self._install_frp()

        # Create config
        self._state.config_file = self._create_config()

        # Build command
        cmd = [str(self._bin_path), "-c", str(self._state.config_file)]

        # Start process
        self._state.process = subprocess.Popen(
          cmd,
          stdout=None,
          stderr=None,
        )

        # Give it a moment to start and check if it failed immediately
        try:
            # Wait for the process to exit or timeout
            stdout_data, stderr_data = self._state.process.communicate(timeout=1.0)

            # If we get here, the process exited within 1 second
            stderr = stderr_data.decode() if stderr_data else ""
            stdout = stdout_data.decode() if stdout_data else ""
            self._cleanup()
            raise RuntimeError(f"FRP process exited immediately: {stderr}. Output: {stdout}")

        except subprocess.TimeoutExpired:
            # Process is still running after 1 second
            pass

      except Exception as e:
        self._cleanup()
        if isinstance(e, RuntimeError):
          raise
        raise RuntimeError(f"Failed to start proxy: {e}") from e

  def stop(self) -> None:
    """Stop the FRP proxy tunnel."""
    with self._state.lock:
      self._cleanup()

  def _cleanup(self) -> None:
    """Internal cleanup method."""
    if self._state.process is not None:
      try:
        self._state.process.terminate()
        try:
          self._state.process.wait(timeout=5.0)
        except subprocess.TimeoutExpired:
          self._state.process.kill()
          self._state.process.wait()
      except Exception:
        pass
      self._state.process = None

    if self._state.config_file is not None:
      try:
        if self._state.config_file.exists():
          self._state.config_file.unlink()
      except Exception:
        pass
      self._state.config_file = None

  @property
  def is_running(self) -> bool:
    """Check if the proxy is currently running."""
    with self._state.lock:
      if self._state.process is None:
        return False
      return self._state.process.poll() is None

  def __enter__(self) -> "FRPProxy":
    """Context manager entry."""
    self.start()
    return self

  def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
    """Context manager exit."""
    self.stop()


# Proxy state persistence (stub for CLI compatibility)
def get_proxy_credentials() -> None:
  """Get credentials for a running proxy (if any).

  Returns:
      None - proxy state persistence not yet implemented.
  """
  return None

File: ts/src/proxy.ts
import { execSync, spawn, type ChildProcess } from "node:child_process";
import {
  createWriteStream,
  existsSync,
  mkdirSync,
  unlinkSync,
  chmodSync,
  writeFileSync,
  rmSync,
} from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { pipeline } from "node:stream/promises";
import { createGunzip } from "node:zlib";
import { Readable } from "node:stream";
import { platform, arch } from "node:process";

const FRP_VERSION = "0.66.0";
const FRP_DIR = join(homedir(), ".smooth", "frp");

export interface ProxyConfig {
  serverUrl: string;
  token: string;
  remotePort?: number;
  sessionId?: string;
}

export class FRPProxy {
  readonly config: ProxyConfig;
  private _process: ChildProcess | null = null;
  private _configFile: string | null = null;
  private _binPath: string | null = null;

  constructor(config: ProxyConfig) {
    this.config = {
      remotePort: 1080,
      sessionId: "default",
      ...config,
    };
  }

  static getPlatformInfo(): { osName: string; arch: string; ext: string } {
    let osName: string;
    if (platform === "darwin") osName = "darwin";
    else if (platform === "win32") osName = "windows";
    else osName = "linux";

    let archName: string;
    if (arch === "x64") archName = "amd64";
    else if (arch === "arm64") archName = "arm64";
    else throw new Error(`Unsupported architecture: ${arch}`);

    const ext = platform === "win32" ? "zip" : "tar.gz";
    return { osName, arch: archName, ext };
  }

  async installFrp(): Promise<string> {
    mkdirSync(FRP_DIR, { recursive: true });

    const { osName, arch: archName, ext } = FRPProxy.getPlatformInfo();
    const binName = osName === "windows" ? "frpc.exe" : "frpc";
    const binPath = join(FRP_DIR, binName);

    if (existsSync(binPath)) return binPath;

    const folderName = `frp_${FRP_VERSION}_${osName}_${archName}`;
    const filename = `${folderName}.${ext}`;
    const url = `https://github.com/fatedier/frp/releases/download/v${FRP_VERSION}/${filename}`;

    const tmpPath = join(FRP_DIR, `tmp_${filename}`);
    const extractDir = join(FRP_DIR, "extract_tmp");

    const maxRetries = 3;
    for (let attempt = 0; attempt < maxRetries; attempt++) {
      try {
        const resp = await fetch(url);
        if (!resp.ok || !resp.body) {
          throw new Error(`HTTP ${resp.status}: ${resp.statusText}`);
        }
        const writeStream = createWriteStream(tmpPath);
        await pipeline(Readable.fromWeb(resp.body as never), writeStream);
        break;
      } catch (e) {
        if (attempt === maxRetries - 1) throw e;
        await new Promise((r) => setTimeout(r, 2 ** attempt * 1000));
      }
    }

    mkdirSync(extractDir, { recursive: true });

    if (ext === "zip") {
      // Use system unzip for simplicity
      execSync(`unzip -o "${tmpPath}" -d "${extractDir}"`, { stdio: "ignore" });
    } else {
      // Use tar for .tar.gz
      execSync(`tar -xzf "${tmpPath}" -C "${extractDir}"`, { stdio: "ignore" });
    }

    const src = join(extractDir, folderName, binName);
    const { renameSync } = await import("node:fs");
    if (existsSync(binPath)) unlinkSync(binPath);
    renameSync(src, binPath);

    // Cleanup
    try {
      unlinkSync(tmpPath);
    } catch {
      /* ignore */
    }
    try {
      rmSync(extractDir, { recursive: true, force: true });
    } catch {
      /* ignore */
    }

    if (osName !== "windows") {
      chmodSync(binPath, 0o755);
    }

    return binPath;
  }

  createConfig(): string {
    mkdirSync(FRP_DIR, { recursive: true });

    const configPath = join(FRP_DIR, `frpc_${this.config.sessionId}.yml`);
    const yaml = `
serverAddr: ${this.config.serverUrl}
serverPort: 443
loginFailExit: false
auth:
  method: token
  token: "${this.config.token}"

log:
  level: "error"

transport:
  protocol: "wss"

proxies:
  - name: "socks5_tunnel_${this.config.sessionId}"
    type: "tcp"
    remotePort: ${this.config.remotePort}
    plugin:
      type: "socks5"
`;
    writeFileSync(configPath, yaml);
    return configPath;
  }

  async start(): Promise<void> {
    if (this._process !== null) {
      throw new Error("Proxy is already running");
    }

    try {
      this._binPath = await this.installFrp();
      this._configFile = this.createConfig();

      this._process = spawn(this._binPath, ["-c", this._configFile], {
        stdio: "ignore",
      });

      // Wait briefly to see if process exits immediately
      await new Promise<void>((resolve, reject) => {
        const timer = setTimeout(() => {
          this._process?.removeAllListeners("exit");
          resolve();
        }, 1000);
        timer.unref();

        this._process!.once("exit", (code) => {
          clearTimeout(timer);
          this._cleanup();
          reject(new Error(`FRP process exited immediately with code ${code}`));
        });
      });
    } catch (e) {
      this._cleanup();
      if (e instanceof Error && e.message.startsWith("FRP process exited")) {
        throw e;
      }
      throw new Error(`Failed to start proxy: ${e}`);
    }
  }

  stop(): void {
    this._cleanup();
  }

  private _cleanup(): void {
    if (this._process) {
      try {
        this._process.kill("SIGTERM");
        // Give it a moment, then force kill
        setTimeout(() => {
          try {
            this._process?.kill("SIGKILL");
          } catch {
            /* ignore */
          }
        }, 5000).unref();
      } catch {
        /* ignore */
      }
      this._process = null;
    }

    if (this._configFile) {
      try {
        if (existsSync(this._configFile)) {
          unlinkSync(this._configFile);
        }
      } catch {
        /* ignore */
      }
      this._configFile = null;
    }
  }

  get isRunning(): boolean {
    if (!this._process) return false;
    return this._process.exitCode === null;
  }
}
